File: day09/solution.go
// Advent of Code 2025 - Day 9: Movie Theater (largest rectangle from tile corners)
package main

import (
	"bufio"
	"fmt"
	"math/bits"
	"os"
	"slices"
	"sort"
	"strconv"
	"strings"
)

// Point represents a red tile position on the grid
type Point struct {
	x, y int
}

// abs returns absolute value (Go doesn't have built-in int abs)
func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

// solve finds the largest rectangle using any two red tiles as opposite corners.
// O(n) approach: the optimal rectangle always involves "extreme" points.
func solve(tiles []Point) int64 {
	if len(tiles) < 2 {
		return 0
	}

	// track extreme points in a single pass
	minXPt, maxXPt := tiles[0], tiles[0]
	minYPt, maxYPt := tiles[0], tiles[0]
	minSumPt, maxSumPt := tiles[0], tiles[0]   // x+y extremes (diagonal)
	minDiffPt, maxDiffPt := tiles[0], tiles[0] // x-y extremes (anti-diagonal)

	for _, p := range tiles {
		// axis-aligned extremes
		if p.x < minXPt.x {
			minXPt = p
		}
		if p.x > maxXPt.x {
			maxXPt = p
		}
		if p.y < minYPt.y {
			minYPt = p
		}
		if p.y > maxYPt.y {
			maxYPt = p
		}

		// diagonal extremes
		sum := p.x + p.y
		diff := p.x - p.y
		if sum < minSumPt.x+minSumPt.y {
			minSumPt = p
		}
		if sum > maxSumPt.x+maxSumPt.y {
			maxSumPt = p
		}
		if diff < minDiffPt.x-minDiffPt.y {
			minDiffPt = p
		}
		if diff > maxDiffPt.x-maxDiffPt.y {
			maxDiffPt = p
		}
	}

	// collect unique candidates using a map
	seen := make(map[Point]bool)
	for _, p := range []Point{minXPt, maxXPt, minYPt, maxYPt, minSumPt, maxSumPt, minDiffPt, maxDiffPt} {
		seen[p] = true
	}

	candidates := make([]Point, 0, len(seen))
	for p := range seen {
		candidates = append(candidates, p)
	}

	// check all pairs among candidates (at most 28 pairs)
	var maxArea int64 = 0
	for i := 0; i < len(candidates); i++ {
		for j := i + 1; j < len(candidates); j++ {
			p1, p2 := candidates[i], candidates[j]
			// +1 because tiles are squares - both corners included
			width := int64(abs(p2.x-p1.x) + 1)
			height := int64(abs(p2.y-p1.y) + 1)
			area := width * height
			if area > maxArea {
				maxArea = area
			}
		}
	}

	return maxArea
}

// VerticalEdge represents an edge going from (x, yLo) to (x, yHi)
type VerticalEdge struct {
	x, yLo, yHi int
}

// solvePart2 finds the largest rectangle with red corners where all tiles are valid.
// Uses sparse tables for O(1) range queries on the polygon's row-wise boundaries.
func solvePart2(tiles []Point) int64 {
	if len(tiles) < 2 {
		return 0
	}

	n := len(tiles)

	// build vertical edges from consecutive red tiles
	verticalEdges := make([]VerticalEdge, 0, n)
	for i := 0; i < n; i++ {
		x1, y1 := tiles[i].x, tiles[i].y
		x2, y2 := tiles[(i+1)%n].x, tiles[(i+1)%n].y
		if x1 == x2 {
			verticalEdges = append(verticalEdges, VerticalEdge{x1, min(y1, y2), max(y1, y2)})
		}
	}

	// collect unique y-values using map, then sort
	ySet := make(map[int]struct{})
	for _, p := range tiles {
		ySet[p.y] = struct{}{}
	}

	allYs := make([]int, 0, len(ySet))
	for y := range ySet {
		allYs = append(allYs, y)
	}
	sort.Ints(allYs) // use stdlib sort

	k := len(allYs)

	// compute left[y] and right[y] for each unique y
	leftArr := make([]int, k)
	rightArr := make([]int, k)

	for i, y := range allYs {
		// collect x's from edges spanning this row + tiles on this row
		xs := make([]int, 0, len(verticalEdges)+len(tiles))
		for _, edge := range verticalEdges {
			if edge.yLo <= y && y <= edge.yHi {
				xs = append(xs, edge.x)
			}
		}
		for _, p := range tiles {
			if p.y == y {
				xs = append(xs, p.x)
			}
		}

		// use slices.Min/Max (Go 1.21+)
		leftArr[i] = slices.Min(xs)
		rightArr[i] = slices.Max(xs)
	}

	// build flattened sparse tables using bits.Len for log2
	logK := bits.Len(uint(k))
	if logK == 0 {
		logK = 1
	}
	tableSize := logK * k
	sparse := make([]int, tableSize*2) // [maxLeft..., minRight...]

	// helper to index: table 0 = maxLeft, table 1 = minRight
	idx := func(table, level, pos int) int {
		return table*tableSize + level*k + pos
	}

	// initialize level 0
	for i := 0; i < k; i++ {
		sparse[idx(0, 0, i)] = leftArr[i]
		sparse[idx(1, 0, i)] = rightArr[i]
	}

	// build higher levels
	for j := 1; j < logK; j++ {
		half := 1 << (j - 1)
		step := half << 1
		if step > k {
			continue
		}
		for i := 0; i <= k-step; i++ {
			sparse[idx(0, j, i)] = max(sparse[idx(0, j-1, i)], sparse[idx(0, j-1, i+half)])
			sparse[idx(1, j, i)] = min(sparse[idx(1, j-1, i)], sparse[idx(1, j-1, i+half)])
		}
	}

	// query function using bits.Len for log2
	query := func(table, lo, hi int) int {
		length := hi - lo + 1
		j := bits.Len(uint(length)) - 1
		idx2 := hi - (1 << j) + 1
		if table == 0 {
			return max(sparse[idx(0, j, lo)], sparse[idx(0, j, idx2)])
		}
		return min(sparse[idx(1, j, lo)], sparse[idx(1, j, idx2)])
	}

	// check all pairs of red tiles
	var maxArea int64 = 0

	for i := 0; i < n; i++ {
		x1, y1 := tiles[i].x, tiles[i].y
		for j := i + 1; j < n; j++ {
			x2, y2 := tiles[j].x, tiles[j].y

			rxLo, rxHi := min(x1, x2), max(x1, x2)
			ryLo, ryHi := min(y1, y2), max(y1, y2)

			potential := int64(rxHi-rxLo+1) * int64(ryHi-ryLo+1)
			if potential <= maxArea {
				continue
			}

			// binary search for y indices (O(log k))
			iyLo := sort.SearchInts(allYs, ryLo)
			iyHi := sort.SearchInts(allYs, ryHi)

			ml := query(0, iyLo, iyHi)
			mr := query(1, iyLo, iyHi)

			if ml <= rxLo && mr >= rxHi {
				maxArea = potential
			}
		}
	}

	return maxArea
}

func loadInput(filename string) ([]Point, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, fmt.Errorf("opening file: %w", err)
	}
	defer file.Close()

	var tiles []Point
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}

		parts := strings.Split(line, ",")
		x, err := strconv.Atoi(parts[0])
		if err != nil {
			return nil, fmt.Errorf("parsing x: %w", err)
		}
		y, err := strconv.Atoi(parts[1])
		if err != nil {
			return nil, fmt.Errorf("parsing y: %w", err)
		}
		tiles = append(tiles, Point{x, y})
	}

	if err := scanner.Err(); err != nil {
		return nil, fmt.Errorf("scanning: %w", err)
	}

	return tiles, nil
}

func main() {
	tiles, err := loadInput("day09/input.txt")
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}

	fmt.Println("--- Part 1 ---")
	part1 := solve(tiles)
	fmt.Printf("Answer: %d\n", part1)

	fmt.Println("--- Part 2 ---")
	part2 := solvePart2(tiles)
	fmt.Printf("Answer: %d\n", part2)
}

File: day09/solution.py
"""Day 9: Movie Theater - finding the largest rectangle between red tiles."""

from collections import deque
from pathlib import Path


def solve_part1(tiles: list[tuple[int, int]]) -> int:
    """Find the largest rectangle using any two red tiles as opposite corners.

    O(n) approach: the optimal rectangle always involves "extreme" points.
    We track 8 extremes (4 axis-aligned + 4 diagonal) and check only those pairs.
    """
    if len(tiles) < 2:
        return 0

    # track extreme points in a single pass
    min_x_pt = max_x_pt = tiles[0]
    min_y_pt = max_y_pt = tiles[0]
    min_sum_pt = max_sum_pt = tiles[0]  # x+y extremes (diagonal)
    min_diff_pt = max_diff_pt = tiles[0]  # x-y extremes (anti-diagonal)

    for x, y in tiles:
        if x < min_x_pt[0]:
            min_x_pt = (x, y)
        if x > max_x_pt[0]:
            max_x_pt = (x, y)
        if y < min_y_pt[1]:
            min_y_pt = (x, y)
        if y > max_y_pt[1]:
            max_y_pt = (x, y)

        s, d = x + y, x - y
        if s < min_sum_pt[0] + min_sum_pt[1]:
            min_sum_pt = (x, y)
        if s > max_sum_pt[0] + max_sum_pt[1]:
            max_sum_pt = (x, y)
        if d < min_diff_pt[0] - min_diff_pt[1]:
            min_diff_pt = (x, y)
        if d > max_diff_pt[0] - max_diff_pt[1]:
            max_diff_pt = (x, y)

    # at most 8 unique candidates
    candidates = list(
        set(
            [
                min_x_pt,
                max_x_pt,
                min_y_pt,
                max_y_pt,
                min_sum_pt,
                max_sum_pt,
                min_diff_pt,
                max_diff_pt,
            ]
        )
    )

    # check all pairs among candidates (at most 28 pairs)
    max_area = 0
    for i in range(len(candidates)):
        for j in range(i + 1, len(candidates)):
            x1, y1 = candidates[i]
            x2, y2 = candidates[j]
            # +1 because tiles are squares - both corners included
            area = (abs(x2 - x1) + 1) * (abs(y2 - y1) + 1)
            max_area = max(max_area, area)

    return max_area


def solve_part2(red_tiles: list[tuple[int, int]]) -> int:
    """Find largest rectangle where corners are red and all tiles are red or green.

    The red tiles form a closed polygon. Green tiles fill the edges and interior.

    Key optimization: For a rectilinear polygon, each row y has a valid x-range
    [left[y], right[y]]. A rectangle is valid iff for all rows in its y-span,
    the rectangle's x-span is within that row's valid range.

    Uses coordinate compression + range queries for efficiency.
    """
    if len(red_tiles) < 2:
        return 0

    n = len(red_tiles)

    # build vertical edges (for computing valid x-ranges per row)
    vertical_edges = []
    for i in range(n):
        x1, y1 = red_tiles[i]
        x2, y2 = red_tiles[(i + 1) % n]
        if x1 == x2:  # vertical edge
            vertical_edges.append((x1, min(y1, y2), max(y1, y2)))

    # get all unique y values and their valid x-ranges
    # for a simple rectilinear polygon, each row has one contiguous valid interval
    all_ys = sorted(set(y for x, y in red_tiles))
    y_to_idx = {y: i for i, y in enumerate(all_ys)}

    # compute left[y] and right[y] for each unique y
    # left[y] = leftmost x in valid region, right[y] = rightmost
    left_arr = []
    right_arr = []

    for y in all_ys:
        # find x-coordinates where vertical edges cross this row
        # for interior points: edges where y_lo < y < y_hi
        # for boundary points: edges where y_lo <= y <= y_hi
        xs = []
        for ex, ey_lo, ey_hi in vertical_edges:
            if ey_lo <= y <= ey_hi:
                xs.append(ex)

        # also collect x's from red tiles on this row (boundary)
        for rx, ry in red_tiles:
            if ry == y:
                xs.append(rx)

        left_arr.append(min(xs))
        right_arr.append(max(xs))

    # build sparse tables for O(1) range-max(left) and range-min(right) queries
    k = len(all_ys)
    log_k = max(1, k.bit_length())

    # sparse table for range maximum of left[]
    max_left = [[0] * k for _ in range(log_k)]
    max_left[0] = left_arr[:]

    for j in range(1, log_k):
        step = 1 << j
        for i in range(k - step + 1):
            max_left[j][i] = max(max_left[j - 1][i], max_left[j - 1][i + (1 << (j - 1))])

    # sparse table for range minimum of right[]
    min_right = [[0] * k for _ in range(log_k)]
    min_right[0] = right_arr[:]

    for j in range(1, log_k):
        step = 1 << j
        for i in range(k - step + 1):
            min_right[j][i] = min(min_right[j - 1][i], min_right[j - 1][i + (1 << (j - 1))])

    def query_max_left(lo, hi):
        """Max of left[lo..hi] inclusive (using indices into all_ys)."""
        length = hi - lo + 1
        j = length.bit_length() - 1
        return max(max_left[j][lo], max_left[j][hi - (1 << j) + 1])

    def query_min_right(lo, hi):
        """Min of right[lo..hi] inclusive."""
        length = hi - lo + 1
        j = length.bit_length() - 1
        return min(min_right[j][lo], min_right[j][hi - (1 << j) + 1])

    # check all pairs of red tiles
    max_area = 0

    for i in range(n):
        x1, y1 = red_tiles[i]
        for j in range(i + 1, n):
            x2, y2 = red_tiles[j]

            # rectangle bounds
            rx_lo, rx_hi = min(x1, x2), max(x1, x2)
            ry_lo, ry_hi = min(y1, y2), max(y1, y2)

            # quick area check
            potential = (rx_hi - rx_lo + 1) * (ry_hi - ry_lo + 1)
            if potential <= max_area:
                continue

            # get y-index range
            iy_lo, iy_hi = y_to_idx[ry_lo], y_to_idx[ry_hi]

            # for rectangle to be valid:
            # max(left[y]) <= rx_lo  (all rows have left boundary <= our left edge)
            # min(right[y]) >= rx_hi (all rows have right boundary >= our right edge)
            ml = query_max_left(iy_lo, iy_hi)
            mr = query_min_right(iy_lo, iy_hi)

            if ml <= rx_lo and mr >= rx_hi:
                max_area = potential

    return max_area


def main():
    text = (Path(__file__).parent / "input.txt").read_text()

    tiles = []
    for line in text.strip().split("\n"):
        x, y = map(int, line.split(","))
        tiles.append((x, y))

    part1 = solve_part1(tiles)
    print(f"Part 1: {part1}")

    part2 = solve_part2(tiles)
    print(f"Part 2: {part2}")


if __name__ == "__main__":
    main()
